ale_linters/cs/mcsc.vim

```diff
@@ -0,0 +1,107 @@
+" general mcs options which are likely to stay constant across
+" source trees like -pkg:dotnet
+let g:ale_cs_mcsc_options = get(g:, 'ale_cs_mcsc_options', '')
+
+" path string pointing the linter to the base path of the
+" source tree to check
+let g:ale_cs_mcsc_source = get(g:,'ale_cs_mcsc_source','')
+
+" list of search paths for additional assemblies to consider
+let g:ale_cs_mcsc_assembly_path = get(g:,'ale_cs_mcsc_assembly_path',[])
+
+" list of assemblies to consider
+let g:ale_cs_mcsc_assemblies = get(g:,'ale_cs_mcsc_assemblies',[])
+function! ale_linters#cs#mcsc#GetCommand(buffer) abort
+
+    " if list of assembly search paths is not empty convert it to
+    " appropriate -lib: parameter of mcs
+    let l:path = ale#Var(a:buffer,'cs_mcsc_assembly_path')
+    if !empty(l:path)
+         let l:path = '-lib:' . join(l:path,',')
+    else
+        let l:path =''
+    endif
+
+    " if list of assemblies to link is not empty convert it to the
+    " appropriate -r: parameter of mcs
+    let l:assemblies = ale#Var(a:buffer,'cs_mcsc_assemblies')
+    if !empty(l:assemblies)
+        let l:assemblies = '-r:' . join(l:assemblies,',')
+    else
+        let l:assemblies =''
+    endif
+
+    " remember the current working directory
+    let l:cwd = expand('#' . a:buffer . ':p:h')
+    " get the base path of the source tree to lint and if
+    " not empty and exists switch working directory to it
+    let l:base = ale#Var(a:buffer,'cs_mcsc_source')
+    if isdirectory(l:base)
+        exe 'cd ' . l:base
+    else
+        let l:base = '.'
+    endif
+
+    " register temporary module target file with ale
+    let l:out = tempname()
+    call ale#engine#ManageFile(a:buffer,l:out)
+
+    " assemble linter command string to be executed by ale
+    " implicitly set -unsafe mcs flag set compilation
+    " target to module (-t:module), direct mcs output to
+    " temporary file (-out)
+    "
+    " TODO mcs would have a built in -recurse flag by which
+    " mcs would be able to recurse through the current directory
+    " tree by it self. In the setup using mcs version 5.2.0.215
+    " it seems not to work reliably, which is why vim built in
+    " glob function is used to collect the list of '*.cs'
+    " files.
+    let l:cmd = 'cd ' . l:base . ';'
+    \    . 'mcs -unsafe'
+    \    . ' ' . ale#Var(a:buffer, 'cs_mcsc_options')
+    \    . ' ' . l:path
+    \    . ' ' . l:assemblies
+    \    . ' -out:' . l:out
+    \    . ' -t:module'
+    \    . ' "' . join(glob('**/*.cs',v:false,v:true),'" "') . '"'
```

Comment:
`glob` will perform very badly on many machines. Can you pass just the one file instead, or something like that?